modules/catalog-compaction/src/main/java/org/apache/ignite/internal/catalog/compaction/CatalogCompactionRunner.java

```diff
@@ -0,0 +1,386 @@
+        AtomicLong minimumRequiredTimeHolder = new AtomicLong(Long.MAX_VALUE);
+        String localMemberName0 = localMemberName;
+
+        for (LogicalNode node : logicalTopologyNodes) {
+            if (localMemberName0.equals(node.name())) {
+                continue;
+            }
+
+            CompletableFuture<NetworkMessage> fut = messagingService.invoke(node, request, ANSWER_TIMEOUT)
+                    .whenComplete((msg, ex) -> {
+                        if (ex != null) {
+                            return;
+                        }
+
+                        long time = ((CatalogMinimumRequiredTimeResponse) msg).timestamp();
+                        long prevTime;
+
+                        // Update minimum timestamp.
+                        do {
+                            prevTime = minimumRequiredTimeHolder.get();
+                        } while (time < prevTime && !minimumRequiredTimeHolder.compareAndSet(prevTime, time));
+                    });
+
+            ackFutures.add(fut);
+        }
+
+        return CompletableFutures.allOf(ackFutures)
+                .thenApply(ignore -> Math.min(minimumRequiredTimeHolder.get(), localMinimumTs));
+    }
+
+    private long determineLocalMinimumRequiredTime() {
+        // TODO https://issues.apache.org/jira/browse/IGNITE-22637 Provide actual minimum required time.
+        return HybridTimestamp.MIN_VALUE.longValue();
+    }
+
+    private CompletableFuture<Set<String>> requiredNodes(Catalog catalog) {
+        HybridTimestamp nowTs = clockService.now();
+        Set<String> required = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+        // TODO https://issues.apache.org/jira/browse/IGNITE-22722 This method needs to be simplified.
+        return collectRequiredNodes(catalog, new ArrayList<>(catalog.tables()).iterator(), required, nowTs);
+    }
+
+    private CompletableFuture<Set<String>> collectRequiredNodes(
+            Catalog catalog,
+            Iterator<CatalogTableDescriptor> tabItr,
+            Set<String> required,
+            HybridTimestamp nowTs
+    ) {
+        if (!tabItr.hasNext()) {
+            return CompletableFuture.completedFuture(required);
+        }
+
+        CatalogTableDescriptor table = tabItr.next();
+        CatalogZoneDescriptor zone = catalog.zone(table.zoneId());
+
+        assert zone != null : table.zoneId();
+
+        List<CompletableFuture<?>> partitionFutures = new ArrayList<>(zone.partitions());
+
+        for (int p = 0; p < zone.partitions(); p++) {
+            ReplicationGroupId replicationGroupId = new TablePartitionId(table.id(), p);
+
+            CompletableFuture<TokenizedAssignments> assignmentsFut = placementDriver.getAssignments(replicationGroupId, nowTs)
+                    .whenComplete((tokenizedAssignments, ex) -> {
+                        // Compaction will be performed even if placement driver doesn't provide table assignments.
+                        if (tokenizedAssignments != null) {
```

Comment:
these nodes are called `required` for a reason. We MUST make sure that we have received minimal time from ALL the nodes hosting tables' partitions. If placement driver doesn't provide assignments (why?), we must interrupt current compaction iteration.

Reply:
> If placement driver doesn't provide assignments (why?),

at least, because assignments for some table in the past can be cleared